core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemUpgradeRequest.java

```diff
@@ -0,0 +1,139 @@
+
+import org.apache.lucene.index.MergeTrigger;
+
+import com.b2international.commons.exceptions.BadRequestException;
+import com.b2international.snowowl.core.branch.Branch;
+import com.b2international.snowowl.core.domain.RepositoryContext;
+import com.b2international.snowowl.core.events.Request;
+import com.b2international.snowowl.core.identity.User;
+import com.b2international.snowowl.core.jobs.JobRequests;
+import com.b2international.snowowl.core.jobs.RemoteJobEntry;
+import com.b2international.snowowl.core.merge.Merge;
+import com.b2international.snowowl.core.repository.RepositoryRequests;
+import com.b2international.snowowl.core.uri.CodeSystemURI;
+import com.b2international.snowowl.core.uri.ResourceURIPathResolver;
+import com.b2international.snowowl.eventbus.IEventBus;
+
+/**
+ * @since 7.13.0 
+ */
+final class CodeSystemUpgradeRequest implements Request<RepositoryContext, String> {
+
+	private static final long serialVersionUID = 1L;
+
+	@NotNull
+	private final CodeSystemURI codeSystem;
+	
+	@NotNull
+	private final CodeSystemURI extensionOf;
+
+	private String codeSystemId;
+	
+	public CodeSystemUpgradeRequest(CodeSystemURI codeSystem, CodeSystemURI extensionOf) {
+		this.codeSystem = codeSystem;
+		this.extensionOf = extensionOf;
+	}
+	
+	void setCodeSystemId(String codeSystemId) {
+		this.codeSystemId = codeSystemId;
+	}
+	
+	@Override
+	public String execute(RepositoryContext context) {
+		if (!codeSystem.isHead()) {
+			throw new BadRequestException("Upgrades can not be started from CodeSystem versions.")
+				.withDeveloperMessage("Use '%s' only instead of '%s'", codeSystem.getCodeSystem(), codeSystem);
+		}
+
+		// get available upgrades 
+		final CodeSystem currentCodeSystem = CodeSystemRequests.prepareGetCodeSystem(codeSystem.getCodeSystem())
+				.setExpand(CodeSystem.Expand.AVAILABLE_UPGRADES + "()")
+				.build()
+				.execute(context);
+		final List<CodeSystemURI> availableUpgrades = currentCodeSystem.getAvailableUpgrades();
+		// report bad request if there are no upgrades available
+		if (availableUpgrades.isEmpty()) {
+			throw new BadRequestException("There are no upgrades available for the CodeSystem '%s'.", codeSystem.getCodeSystem());
+		}
+		
+		// or the selected extensionOf version is not present as a valid available upgrade
+		if (!availableUpgrades.contains(extensionOf)) {
+			throw new BadRequestException("Upgrades can only be performed to the next available version dependency.")
+				.withDeveloperMessage("Use '%s/<VERSION_ID>', where <VERSION_ID> is one of: '%s'", extensionOf.getCodeSystem(), availableUpgrades);
+		}
+		
+		// auto-generate the codeSystemId if not provided
+		// auto-generated upgrade IDs consist of the original CodeSystem's name and the new extensionOf dependency's URI
+		final String upgradeCodeSystemId;
+		if (codeSystemId == null) {
+			upgradeCodeSystemId = String.format("%s-%s-UPGRADE", codeSystem.getCodeSystem(), extensionOf.getPath() /*versionId*/); 
```

Comment:
What happens if I upgrade a code system which has already been upgraded once?

Reply:
An upgrade is essentially a branch, so if you start it again you will get a 409 Conflict status indicating that the branch already exists and you just cannot create another Branch/CodeSystem that represents the upgrade. Theoretically, it is possible to have multiple ongoing upgrades to different versions, but we will consider restricting those in the upcoming 7.15.0 release.

Reply:
I was thinking about whether this expression would create `SNOMEDCT-UK-SNOMEDCT/20200131-UPGRADE-SNOMEDCT/20200731-UPGRADE`-kind of short names if I upgrade to 01-31, then to 07-31.